README.md
# cerberus-core #

## Summary ##

This tool aims to help ISP manage abuse on their network by providing
a set of tools such as an email parser and classification, a ticketing
system, an API and an UX for operators (with cerberus-ux project).

This tool can be easily expanded by implementing new `python-rq`jobs.


## Try it ##

### The easy way (Docker) ###

To test the poc (not persistent), run (tested with version 1.9.1):

    $ docker build -t cerberus .

Then, you just need to run it with your abuse email inbox settings (over IMAPS):

    $ docker run -e EMAIL_HOST=mx.domain.com -e EMAIL_PORT=993 -e EMAIL_LOGIN=login@domain.com -e EMAIL_PASS=pass -p 6060:6060 -t cerberus

and browse `http://127.0.0.1:6060` with default login admin:admin , and voila

![cerberus](https://raw.githubusercontent.com/ovh/cerberus-core/devel/doc/source/cerberus.gif?token=AQVCyzi29XLlKOim8_VoCJjYEWyoY1Gaks5Wx0qqwA%3D%3D)

If something goes wrong, enter the docker and check the logs in `/home/cerberus/cerberus-core/*.log`

### The manual way ###

This project is mainly based on **Django**, **Flask**, **RQ** and **Rq-Scheduler**. To setup cerberus-core, you'll need:

 * A Linux environment
 * Python 2.7+
 * Packages `python-dev` and `python-pip`
 * A PostgreSQL database (9.4.x or greater)
 * An HTTP Server for cerberus-ux (not included in this project)
 * A Redis server
 * A MX supporting IMAPS
 * A scheduler (cron, supervisor, ...)

When all of these requirements are met, you can install the tool:

 1. Download the zipfile or checkout the sources.
 2. install python dependencies (`apt-get install python-dev python-pip`)
 3. Run `pip install -r requirements/common.txt` in order to setup dependencies

You should **take a look at docker directory** for further informations.

## Running ##

Start the scheduler:

    $ rqscheduler &

You can now insert theses entries in your favorite scheduler:

 * Schedule `rqcheduler`, `event/email_fetcher.py`, `worker/worker.py` and `uwsgi --http-socket 8080 api/uwsgi.ini` to be run as a daemon.
 * Schedule `event/workflow.py` to be run every minute.
 * Schedule `event/stats.py` to be run every quarter hour.

The whole project use `python-rq` and `rq-scheduler`. You can see jobs status with:

    $ rq-info

## Concept ##

### Plaintiff ###

(Law) A a person who brings a civil action in a court (aka claimant).

### Defendant ###

(Law) A person or entity against whom an action or claim is brought in a court of law.
It is one of your customer who is suspected of illegal activities on his service(s).

### Service ###

A Defendant has suscribed to one or more services(s) (product(s)) your company offers. It can be ADSL, hosting, email services ...

### Provider ###

The source of the email. It can be directly the plaintiff or an representative third party.

### Report ###

One Provider + one Category. If a defendant is identified, the report is linked to a defendant, a service and contains item(s).

### Ticket ###

One or more reports + one Category. It cans bel linked to a defendant/service, so all reports themselves linked to this defendant/service.

### Item ###
        
Fraudulent URL/IP/FQDN found in an email.

### Action ###

An action on a customer's service (suspend, shutdown, breach of contract).

## Workflow ##

An email can generates one or more reports (if multiple defendants are identified). Just one if no defendant is identified.
These reports can be attached to existing ticket or create one if the provider is "trusted".
So tickets can have multiple reports/providers.

So **not all reports are attached to tickets**. It's important, operators **process tickets, not reports**. Reports add weight to tickets.

All the effective jobs are done with ticket: customer interaction (emails), action on service ...

There's an specific/automatic workflow for **phishing** ticket/report described in `doc/source/phishing.png`.
A "Phishing-trusted reporter" is a provider with an `apiKey` (see `abuse/models.py`).

You can see full cerberus models's relationship in `doc/source/models.png`.

## Configuration ##

### Project structure ###

 * `abuse`: Django app models description.
 * `adapters`: Abstract classes providing way to implement core functions.
 * `api`: cerberus-core API for cerberus-ux.
 * `default`: default implementation of adapters abtract classes.
 * `doc`: documentation.
 * `docker`: files needed to build docker.
 * `event`: event-based jobs to be consume by python-rq workers.
 * `factory`: the factory for adapters implementations.
 * `requirements`: pip requirements.
 * `tests`: unit tests.
 * `utils`: utils functions for worker and API.
 * `worker`: python-rq workers.

### General settings ###

For security purpose, most of setting values must be defined as VARENV. So, to configure the tool, you just have to export following environment variables:

 * EMAIL_STORAGE_DIR: The folder where to store fetched and sent emails
 * MAGIC_SMTP_HEADER: King of tag in SMTP header defining e-mail source (see next section)
 * API_HOST: The IP address of cerberus-core API
 * API_PORT: The TCP port of cerberus-core API
 * EMAIL_HOST: IP or domain of your MX
 * EMAIL_PORT: The TCP port of your MX (IMAPS)
 * EMAIL_LOGIN: Username used to poll incoming e-mails
 * EMAIL_PASS: Password for previous username
 * REDIS_HOST: The IP address of the Redis Server
 * REDIS_PORT: The TCP port of the Redis Server
 * SECRET_KEY: The Django secret key
 * PG_NAME: The name of the PostgreSQL db
 * PG_USER: The username to use for previous db
 * PG_PASS: The password of this user
 * PG_HOST: The IP address of PostgreSQL db
 * PG_PORT: The TCP port of PostgreSQL db

In `settings.py` you can see how this different varenv are used + other settings description. **Really, you should edit this file.**

### A "trusted" email ? ###

As you can see, a required varenv is called MAGIC_SMTP_HEADER.
Since everybody is able to send an email pretending being Microsoft, NSA or whatever,
it's important to check the identity of the mail send with a well-kept secret.
The easier way to do so is to assign a single e-mail address per organization you keep secret.

Using the MX features, the mail must be validated and tagged with the previous MAGIC_SMTP_HEADER.
We recommend to keep this header name secret too. No matter the value of this header, if it's present, the email is trusted.

### Network IPs ###

When you're done, with general config, the tool is almost ready to be run. You now have to tell the IPs your network is made of.
Everything you have to do is opening the file utils/ips.py and add your CIDRs (only one per line).

### Implementing your own core functions ###

You'll maybe see in the file `settings.py`, there is a way to provide your implementation(s).
By default, a very basic implementation is provided for each adapter. (see abstract documentation)

Required adapters implementations are:

 * adapters.services.storage.abstract.StorageServiceBase
 * adapters.dao.customer.abstract.CustomerDaoBase
 * adapters.services.phishing.abstract.PhishingServiceBase
 * adapters.services.mailer.abstract.MailerServiceBase
 * adapters.services.action.impl.ActionServiceBase

Optional, but usefull, are:

 * adapters.services.kpi.abstract.KPIServiceBase
 * adapters.services.search.abstract.SearchServiceBase
 * adapters.dao.reputation.impl.ReputationDaoBase

You can code your own service(s) by implementing `adapters.services.*.abstract.*` and your own DAO by implementing `adapters.dao.*.abstract.*.`
Then, tell cerberus-core to use this implementations by editing the property CUSTOM_IMPLEMENTATIONS.

You can find implementation and expected return value in `default.adapters`.

### Add features ###

You can easily add event in `event/` and there associated functions in `worker/`
You can add template for provider in `worker/parsing/templates`. If your template is well formatted, it will be added to parser regexp.

## API ##

Once everything is running, you can start using the API. By default, it's listening to the port 8080. Full endpoints description are available in documentation

**Be careful, this is not a RESTFul API.** The main goal of this API is to interface DB with `ovh/cerberus-ux` project.

## Documentation ##

You can build the full documentation with:

    $ sphinx-build -b html doc/source doc/build


CHANGELOG
2016-10-06 : v1.1

* add(test): for new defendant models (and fix)
* feaf(): api and worker func for unblock url
* feat(): ABUSE-721 Utiliser différentes priorités par rq:queue
* feat(): ABUSE-803
* feat(): ABUSE-808
* feat(): ABUSE-822
* feat(adapters.phishing): unblock url
* feat(adapters.services.mailer): add attachments on send_email
* feat(): add ticket's answer attachments to mailerservice
* feat(): add worker func/api endpoint for threshold
* feat(api): ABUSE-819
* feat(api): add cache on some endpoints
* feat(api): add onlyUnassigned filters on todo tickets
* feat(api): add ticket attachments
* feat(api): attachments in emails
* feat(api): avoid double ticket creation for "ToValidate" reports
* feat(api): cancel async jobs when ticket is closed
* feat(api): change ticket search filters conditions
* feat(api): get timeline
* feat(api): huge decorator refactor
* feat(api): improve perf
* feat(api): limitedOpen scheduling algo avoid brand new customer
* feat(api): list created mass contact campaign
* feat(api): mass contact controller
* feat(api): mass-contact now with data validation
* feat(api): more cached endpoints
* feat(api): new run level ok for api
* feat(api): refactor
* feat(api): update prioritized tickets alog, critical before all others
* feat(api): user's are now priorized in 'todo' tickets
* feat(api): whois-like
* feat(api/worker): use django templating in email
* feat(): check mass contact campaign result
* feat(*): data validation with voluptuous for CustomerDaoBase
* feat(event): add report threshold event producer
* feat(): first implementation without tests
* feat(): first try
* feat(): get_http_headers
* feat(): new defendant model on API
* feat(*): new defendant models for workers
* feat(): new ticket history management
* feat(*): new workflow for trusted copyright providers
* feat(): now asynchronously update defendant infos
* feat(): now using timeout for rq_scheduler jobs
* feat(): now worker/API use same function to retreive item infos
* feat(parsing): add blacklist of cidr
* feat(parsing): implement blacklisted provider check
* feat(parsing): update ACNS regexp
* feat(phishing): add phishingGrade and score to get_screenshot
* feat(phishing.timeout): closing if all down too
* feat(): refactor
* feat(): refactor 2
* feat(): role edition on users and presets
* feat(): specific workflow when phishing is clearly identified
* feat(): store mass-contact campaign details
* feat(test): add new test for automated phishing
* feat(test): fix test for new models
* feat(): tests
* feat(): ticket answer generic workflow
* feat(ticket): move ticket pause/unpause code from api to worker
* feat(): update model, ABUSE-777
* feat(): Update "priorized tickets" algo
* feat(): user list instead of set for ReportWorkflowHookFactory to respect sequence
* feat(utils): replace reverse with NXDOMAIN is now optionnal
* feat(worker): ABUSE-817
* feat(worker): add items to proof on phishtocheck validation
* feat(worker): add log
* feat(worker): add reparse_validated function
* feat(worker): cancel ticket timeout when on ticket answer
* feat(worker): export common functions
* feat(worker): implement copyright/acns workflow
* feat(worker): move report workflows
* feat(worker): new generic report workflow factory
* feat(worker.parsing.template): now can specify if fallback to other templates or not
* feat(worker): split functions on phishing hook
* feat(worker): threshold functions
* feat(worker): unassigned ticket if parameter specified in user.role config
* feat(worker): worker func for mass contact
* feat(workflow): change fts value
* fix(api): 500 when cancelling a job on a ticket with no ServiceAction
* fix(api): add 'ActionError' ticket to myTicketsTodoCount in toolbar
* fix(api): autogenerate codename when creating email template
* fix(api): avoir changing report status without check
* fix(api): bulk
* fix(api): bulk actions on ticket
* fix(api): defendant infos refresh
* fix(api): defendant top20
* fix(api): get defendant from PCS url
* fix(api): improve bulk perf
* fix(api): != instead of is not
* fix(api): interact authorizations checks
* fix(api): limitedOpen algo ..
* fix(api): limitedOpen scheduling algo
* fix(api): limited scheduling algo
* fix(api): multiple same items on report/ticket
* fix(api): now can set no defaultCategory on a provider
* fix(api): now do not cancel all previous pending ServiceActionJobs all when scheduling a new job
* fix(api): overrid defendant id with details id
* fix(api): rewrite top20 db request
* fix(api): set larger timeout for aysnc service action tasks
* fix(api): simplify search
* fix(api.ticket): check if body is correct in interact
* fix(api): title email category in interact to match model
* fix(api): typo
* fix(api): use english language by default pour Plaintiff templates
* fix(api): utf-8 attachments
* fix(api): wrong idea to lower dict key
* fix(): catch more db connection error exceptions
* fix(doc): build
* fix(email_fetcher): unecessary storage re-opening
* fix(mailerdaemon): typo
* fix(): now creates closed ticket if report is not validated
* fix(): now really use redis settings
* fix(): now use memory cache for testing
* fix(parsing): bad url cleaning
* fix(parsing): netaddr lib does not like ip like 038.140.010.024
* fix(parsing): now try to get inline attachments
* fix(parsing): provider's log
* fix(parsing): special case for csv-like (comma separated) urls in report
* fix(parsing): unusual way to arrange "to" recipients
* fix(): pep8 compliance
* fix(phishing): email was sent before items attachments
* fix(): queue name
* fix(readme): cerberus ux img
* fix(readme): link to gif
* fix(): redis settings
* fix(): SonarQube compliance
* fix(test): indentation
* fix(test): mock rq
* fix(test): ok, now really mocked ..
* fix(test): pep8
* fix(): tests
* fix(test): utils
* fix(worker): add proof for copyright workflow
* fix(worker): avoid create multiple 'ticket.timeout' jobs on PhishToCheck validation

settings.py
# -*- coding: utf-8 -*-
#
# Copyright (C) 2015-2016, OVH SAS
#
# This file is part of Cerberus-core.
#
# Cerberus-core is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""
    Settings for Cerberus
"""


import os

# ################################ Cerberus Settings ##################################

# Main Cerberus Configuration
#
# Do not forget EMAIL_STORAGE_DIR ENV (default '/tmp/cerberus_storage_dir')
#
# What's 'magic_smtp_header' ? : A header your SMTP header can add to email depending on sender/mailbox/country etc ...
#
#   It's used to know if an email can be trusted

GENERAL_CONFIG = {
    'customer_dao_datetime_format': '%Y-%m-%d %X',
    'email_storage_dir': os.getenv('EMAIL_STORAGE_DIR', '/tmp/cerberus_storage'),
    'cerberus_emails_db': 'cerberus_emails.db',
    'bot_user': 'abuse.robot',
    'phishing': {
        'wait': 172800,
        'up_threshold': 0,
        'down_threshold': 75,
    },
    'copyright': {
        'wait': 172800,
        'acns_patterns': ('www.acns.net/ACNS',),
        'trusted_copyright_providers': [],
    },
    'magic_smtp_header': os.getenv('MAGIC_SMTP_HEADER', 'X-MAGIC-SMTP-HEADER-TO-IDENTIFY-TRUSTED-PROVIDER'),
    'report_timeout': 30,
    'mailer_daemon_patterns': [
        'mailer-daemon',
        '@provider.com'
    ],
}

# RQ Queues
#

QUEUE = {
    'default': {
        'name': 'default',
        'default_timeout': 86400,
    },
    'email': {
        'name': 'email',
        'default_timeout': 86400,
    },
    'kpi': {
        'name': 'kpi',
        'default_timeout': 1800,
    },
}

# API Config (via ENV)
#
# What's FORWARDED_HOST ? For security reason, if you use different host for Cerberus-UX and Cerberus-API (not required)
#
API = {
    'host': os.getenv('API_HOST', '127.0.0.1'),
    'port': os.getenv('API_PORT', 6060),
    'forwarded_host': os.getenv('FORWARDED_HOST'),
    'use_cache': True,
    'cache_engine': 'redis',  # 'redis' or 'memory'
}

# Defined here your customer adapters implementations. For testing you cans use provided default impl
#
# Required are :
#       adapters.services.storage
#       adapters.dao.customer
#       adapters.services.phishing
#       adapters.services.mailer
#       adapters.services.action

CUSTOM_IMPLEMENTATIONS = (
    'default.adapters.services.storage.impl.FilesystemStorageService',
    'default.adapters.dao.customer.impl.DefaultCustomerDao',
    'default.adapters.services.phishing.impl.DefaultPhishingService',
    'default.adapters.services.mailer.impl.DefaultMailerService',
    'default.adapters.services.action.impl.DefaultActionService',
)

# /!\ Order matter
CUSTOM_REPORT_WORKFLOWS = (
    'worker.workflows.report.phishing.PhishingReportWorkflow',
    'worker.workflows.report.copyright.CopyrightReportWorkflow',
    'worker.workflows.report.acns.AcnsReportWorkflow',
)

# /!\ Order matter
CUSTOM_TICKET_ANSWER_WORKFLOWS = (
    'worker.workflows.ticket.mailerdaemon.MailerDaemonWorkflow',
    'worker.workflows.ticket.customeranswer.CustomerAnswerWorkflow',
    'worker.workflows.ticket.default.DefaultAnswerWorkflow',
)

CUSTOM_SCHEDULING_ALGORITHMS = (
    'api.controllers.scheduling.global.GlobalSchedulingAlgorithm',
    'api.controllers.scheduling.limitedOpen.LimitedOpenSchedulingAlgorithm',
)

# Cerberus use a lot of 'tags',required are here with their mapping to default provided data
#
TAGS = {
    'autoarchive': 'distrust:0:autoarchive',
    'attach_only': 'distrust:1:attach_if_exists',
    'no_phishtocheck': 'distrust:2:no_phishtocheck',
    'no_autoack': 'never_auto_ack',
    'phishing_autoblocked': 'phishing:autoblocked',
    'phishing_autoclosed': 'phishing:autoclosed',
    'copyright_autoclosed': 'copyright:autoclosed',
    'phishing_autoreopen': 'phishing:autoreopened',
    'phishing_toblock': 'phishing:toblock',
}

# Cerberus use a lot of 'codenames',required are here with their mapping to default provided data
#
CODENAMES = {
    'ack_received': 'ack_report_received',
    'case_closed': 'case_closed',
    'first_alert': 'first_alert',
    'fixed_customer': 'fixed_by_customer',
    'fixed': 'fixed_by_isp',
    'forward_acns': 'forward_acns',
    'no_more_content': 'no_more_content',
    'phishing_blocked': 'phishing_blocked',
    'service_blocked': 'phishing_service_blocked',
    'ticket_closed': 'ticket_closed',
    'not_managed_ip': 'not_ovh_ip',
    'invalid': 'invalid',
}

# Above all, Cerberus is basically an abuse's email fetcher
#
# This settings are required (using IMAP)
#
# cerberus_email: The email to use for sending email to defendant
#
# cerberus_re: Regex to know if the email is an anwser to a Cerberus Ticket
#
EMAIL_FETCHER = {
    'host': os.getenv('EMAIL_HOST'),
    'port': os.getenv('EMAIL_PORT', 993),
    'login': os.getenv('EMAIL_LOGIN'),
    'pass': os.getenv('EMAIL_PASS'),
    'cerberus_email': 'ticket+%s.%s@company.com',
    'cerberus_re': r'ticket\+(\w+).(\w+)@company.com',
}

# Parsing settings
#
# 'domain_to ignore': Very usefull to ignore wrong/invalid/unwanted domain when parsing email content
#
# 'fqdn_re': of course trying to identify a FQDN in an text can generate a huge amount of false positive
#           maybe you have a regexp to identify YOUR managed FQDN (i.e: endswith domain.com)
PARSING = {
    'providers_to_ignore': [
        'bad@provider.com',
        'invitations@linkedin.com',
        'messages-noreply@linkedin.com',
    ],
    'networks_to_ignore': [
        '0.0.0.0/8',
        '224.0.0/4',
    ],
    'domain_to_ignore': [
        'www.yourcompany.com',
    ],
    'fqdn_re': r'(.*\.yourcompany\.com)',
}

# Redis needed for Worker/API
#
REDIS = {
    'host': os.getenv('REDIS_HOST', '127.0.0.1'),
    'port': os.getenv('REDIS_PORT', 6379),
}

# You can use 'stderr' or 'gelf' or both
#
LOG = {
    'handlers': ['stderr'],  # or ['gelf'] or ['stderr', 'gelf']

    'gelf': {
        'host': os.getenv('GELF_HOST', '0.0.0.0'),
        'port': int(os.getenv('GELF_PORT', 12345)),
        'static_fields': {
            os.getenv('GELF_FIELD', 'test_field'): os.getenv('GELF_VALUE', 'test_value'),
        }
    },
}

# ################################ Django Settings ##################################

# You need to edit this !!

SECRET_KEY = os.getenv('SECRET_KEY', 'test')

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.postgresql_psycopg2',
        'NAME': os.getenv('PG_NAME', 'abuse'),
        'USER': os.getenv('PG_USER', 'cerberus'),
        'PASSWORD': os.getenv('PG_PASS', 'cerberus'),
        'HOST': os.getenv('PG_HOST', '127.0.0.1'),
        'PORT': os.getenv('PG_PORT', 5432),
    }
}

TIME_ZONE = 'Europe/Paris'

MIDDLEWARE_CLASSES = (
    'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
)

INSTALLED_APPS = (
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'abuse',
)
